File: src/org/angularjs/AngularJSCustomAttributeDescriptorsProvider.java
package org.angularjs;

import com.intellij.psi.xml.XmlTag;
import com.intellij.xml.XmlAttributeDescriptor;
import com.intellij.xml.XmlAttributeDescriptorsProvider;
import com.intellij.xml.impl.schema.AnyXmlAttributeDescriptor;
import com.intellij.xml.util.HtmlUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * @author John Lindquist
 */
public class AngularJSCustomAttributeDescriptorsProvider implements XmlAttributeDescriptorsProvider {
    @Override
    public XmlAttributeDescriptor[] getAttributeDescriptors(XmlTag tag) {
        if (tag == null) {
            return XmlAttributeDescriptor.EMPTY;
        }
        final List<XmlAttributeDescriptor> result = new ArrayList<XmlAttributeDescriptor>();
        //todo: refactor to loading a file
        //todo: explore making the loaded file a DTD/schema that would also make these "valid" attributes for inspections
        String[] directiveNames = {
                "click",
                "dblclick",
                "mousedown",
                "mouseup",
                "mouseover",
                "mouseout",
                "mousemove",
                "mouseenter",
                "mouseleave",
                "app",
                "bind",
                "bind-html-unsafe",
                "bind-template",
                "class",
                "class-even",
                "class-odd",
                "cloak",
                "controller",
                "form",
                "hide",
                "href",
                "include",
                "init",
                "non-bindable",
                "pluralize",
                "repeat",
                "show",
                "submit",
                "style",
                "switch",
                "switch-when",
                "switch-default",
                "options",
                "view",
                "transclude",
                "model",
                "list",
                "change",
                "value",
                "required",

                "checked",
                "csp",
                "multiple",
                "readonly",
                "src"
        };

        for (String directiveName : directiveNames) {
            directiveName = "ng-" + directiveName;
            result.add(new AnyXmlAttributeDescriptor(directiveName));
        }
        result.add(new AnyXmlAttributeDescriptor("required"));

        return result.toArray(new XmlAttributeDescriptor[result.size()]);
    }

    @Override
    public XmlAttributeDescriptor getAttributeDescriptor(String attributeName, XmlTag context) {
        if (context != null) {
            return new AnyXmlAttributeDescriptor(attributeName);
        }
        return null;
    }

}

File: src/org/angularjs/AngularJSCustomAttributeDescriptorsProvider.kt
package org.angularjs

import com.intellij.psi.xml.XmlTag
import com.intellij.xml.XmlAttributeDescriptor
import com.intellij.xml.XmlAttributeDescriptorsProvider
import com.intellij.xml.impl.schema.AnyXmlAttributeDescriptor
import java.util.ArrayList
import com.intellij.xml.impl.dom.DomAttributeXmlDescriptor
import com.intellij.util.xml.DomElement
import com.intellij.util.xml.DomManager
import com.intellij.util.xml.reflect.DomAttributeChildDescription
import java.util.HashMap
import kotlin.properties.Delegates

public open class AngularJSCustomAttributeDescriptorsProvider(): XmlAttributeDescriptorsProvider {
    var angularjs:AngularJS? = null

    public override fun getAttributeDescriptors(tag: XmlTag?): Array<XmlAttributeDescriptor>? {
        if(angularjs == null) angularjs = AngularJS.getInstance(tag?.getProject())

        if (tag == null)
        {
            return XmlAttributeDescriptor.EMPTY
        }

        return angularjs?.attrArray
    }
    public override fun getAttributeDescriptor(attributeName: String?, context: XmlTag?): XmlAttributeDescriptor? {
        if (context != null)
        {
            val descriptor = angularjs?.attrLookup?.get(attributeName)
            if(descriptor == null) return null
            return descriptor
        }

        return null
    }


}
